Update keylogger.c

Ixiodor opened the pull request.

Ixiodor committed:
Update keylogger.c

gianlucag commented:
quoto [Wikipedia](https://en.wikipedia.org/wiki/Message_loop_in_Microsoft_Windows): 

> "It is conventional for the event loop to call TranslateMessage on each message which can translate virtual keystrokes into strings. Calling TranslateMessage is not technically required, but problems can result if it is not called. The message loop must call DispatchMessage."

Ixiodor commented:
Quoto [Microsoft](https://msdn.microsoft.com/it-it/library/windows/desktop/ms644955(v=vs.85).aspx):

> Translates virtual-key messages into character messages. The character messages are posted to the calling thread's message queue, to be read the next time the thread calls the GetMessage or PeekMessage function.

Tu non lo fai, questo serve per l'evento WM_CHAR, la chiamata hook arriva moltissimo prima che venga inoltrato il WM_CHAR, lo puoi verificare comunque da te vedendo che non verrà mai eseguto il TranslateMessage. Il TranslateMessage  serve per la window pump classica e non quella in hook.

gianlucag commented:
ah si, giusto, ma allora perche' non eliminiamo del tutto il thread tenuto in vita dal while() loop? E' necessario secondo te?

Ixiodor committed:
Update keylogger.c

Infatti puoi fare così, funziona uguale. La GetMessage non ritorna perchè non avrà messaggi da darti in effetti,l'unico che avrai sarà quello del WM_QUIT perchè lo manderai tu.

Ixiodor commented:
il while puoi toglierlo completamente, puoi lasciare solo "GetMessage(&msg, NULL, 0, 0);"
Ti ho fatto un commit togliendo il while. Se avessi avuto l'esigenza di dover togliere e mettere l'hook, avresti sbattuto contro un deadlock.

gianlucag commented:
mi pare un'ottima soluzione, do un'occhiata stasera e poi vado di merge

ci sarebbero anche altri punti da sistemare, ad esempio la creazione della png dai dati raw ABGR della bitmap richiede un loop preliminare con complessita' quadratica non bellissimo, serve per "rigirare" le quadruple in RGBA (da ABGR) e flippare la bitmap in verticale
